ONSdigital/rasrm-acceptance-tests
# make enrolment code acceptance tests standalone

# Motivation and Context
Some Acceptance tests rely on data state set by previously executed tests. As each feature is run alphabetically, if the feature file name changes, tests could fail. This PR is an attempt to establish a common pattern using enrolment code related tests, that can be applied to make every test fully independent. Also, Survey/Collection Execise data used should make it easier to trace what Feature/Scenario created it.

# What has changed
A new STANDALONE environment variable to put the tests into standalone mode (currently ONLY prevents the 4 collection exercises + repondent 49900000001 running at startup).

Enrolment code Scenarios "Display an unused active code", "Make a request for a new code", "Frontstage can see the survey they are enrolling in" & "Frontstage user can create an account" have a new @standalone tag.

A new run_tags.py file to allow tests to run in isolation.

The 1st step of each Scenario calls a method for setting up it's data - i.e setup_survey_enrolment_for_test(context)

	This calls a generic survey_utilities.setup_survey_for_test(context) method which after creating a unique ID (used as RU Reference) -

		Creates a new Survey with -
			Survey name = 'TEST-' + feature name + UTC creation timestamp
			Short name = RU Reference
			Survey Type defaults to = 'Business'
			Survey Ref defaults to a randomly generated number between 1001 and 9999
			Survey Legal Basis defaults to 'STA1947' (Statistics of Trade Act 1947)

		Creates a new Survey Collection Exercise with - 
			User Description = 'TEST-' + scenario name (may need compacting)
			A Respondent (if required) with email generated from RU Reference e.g 53621951874@53621951874.com

	Scenario specific data is created

# How to test?
Prove all tests run as before by running usual "make wait_for_services setup acceptance_tests", all should pass and 4 additonal Surveys (prefixed with "TEST-") should be visible on "View list of business surveys" page 

Only standalone tests run -
	from a terminal in the rasrm-acceptance-tests directory, run -
		pipenv shell
		export STANDALONE='True'
		python run_tags.py standalone

		only the 4 enrolment code tests should run and pass

A single test runs -
	Add @wip to a Scenario

	from a terminal in the rasrm-acceptance-tests directory, run -
		pipenv shell
		export STANDALONE='True'
		python run_tags.py wip

		only this test should run and pass

Combinations of tagged tests run -
	Add a mixture of @standalone & @wip tags to Scenarios

	from a terminal in the rasrm-acceptance-tests directory, run -
		pipenv shell
		export STANDALONE='True'

		Scenarios with both (AND) tags run -
			python run_tags.py standalone wip

		Scenarios with either (OR) tags run -
			python run_tags.py standalone,wip

In all the above, non-independant test data (4 collection exercises & 49900000001 Respondent) should not run

# Links
Test in conjunction with https://github.com/ONSdigital/response-operations-ui/pull/241
